### PR TITLE
Issue #71: Various Gremlin query patterns throw NPE when query.batch=true

### PR DESCRIPTION
**Pull request by twilmes**
Storage adapter tests were added to exercise all backends in `query.batch` mode.
The `JanusGraphLocalQueryOptimizerStrategy` was updated to disable multiquery on Janus specific steps that are nested within `RepeatStep`, `MatchStep`, and `BranchSteps`.

These steps cause difficulties with the barrier step nature of the batched query multiquery enabled steps. We should still be able to get at least some level of batched backend retrieval working with these but it will be a more involved effort. The intent of this PR is to un-break the `query.batch` option for users and to allow them to benefit from the decreased latencies in other common query patterns and to lay down a foundation for testing out further multiquery optimizations.

-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there an issue associated with this PR? Is it referenced in the commit message?
- [x] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you written and/or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [ ] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?
- [ ] If this PR is a documentation-only change, have you added a `[skip ci]`
  tag to the first line of your commit message to avoid spending CPU cycles in
  Travis CI when no code, tests, or build configuration are modified?

### Note:
Please ensure that once the PR is submitted, you check Travis CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: janusgraph-hbase-parent/janusgraph-hbase-core/src/test/java/org/janusgraph/blueprints/HBaseMultiQueryGraphProvider.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+import org.janusgraph.HBaseStorageSetup;
+import org.janusgraph.diskstorage.configuration.ModifiableConfiguration;
+import org.apache.commons.configuration.Configuration;
+import org.apache.tinkerpop.gremlin.structure.Graph;
+import org.janusgraph.graphdb.configuration.GraphDatabaseConfiguration;
+
+import java.io.IOException;
+
+/**
+ * @author Matthias Broecheler (me@matthiasb.com)
```
Comment (pluradj) [inline janusgraph-hbase-parent/janusgraph-hbase-core/src/test/java/org/janusgraph/blueprints/HBaseMultiQueryGraphProvider.java]: cut & paste error?

File: janusgraph-test/src/test/java/org/janusgraph/blueprints/InMemoryMultiQueryGraphProvider.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+import org.janusgraph.StorageSetup;
+import org.janusgraph.diskstorage.configuration.ModifiableConfiguration;
+
+/**
+ * @author Ted Wilmes (twilmes@gmail.com)
+ */
+public class InMemoryMultiQueryGraphProvider extends AbstractJanusGraphProvider {
+    @Override
+    public ModifiableConfiguration getJanusGraphConfiguration(String graphName, Class<?> test, String testMethodName) {
+        return StorageSetup.getInMemoryConfiguration();
```
Comment (pluradj) [inline janusgraph-test/src/test/java/org/janusgraph/blueprints/InMemoryMultiQueryGraphProvider.java]: missing `.set(GraphDatabaseConfiguration.USE_MULTIQUERY, true)`

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/JanusGraphLocalQueryOptimizerStrategy.java [hunk 3]
```diff
@@ -149,12 +162,24 @@ private static void unfoldLocalTraversal(final Traversal.Admin<?, ?> traversal,
             vstep.setTraversal(traversal);
             TraversalHelper.replaceStep(localStep, vstep, traversal);
 
-            if (useMultiQuery) {
+            if (useMultiQuery) {// && !(isChildOf(vstep, MULTIQUERY_INCOMPATIBLE_STEPS))) {
```
Comment (pluradj) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/JanusGraphLocalQueryOptimizerStrategy.java]: Why is this code commented out?

File: janusgraph-test/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java [hunk 4]
```diff
@@ -3518,6 +3518,10 @@ public void testTinkerPopOptimizationStrategies() {
         TraversalMetrics metrics;
         GraphTraversalSource gts = graph.traversal();
 
+        clopen(option(USE_MULTIQUERY), true);
+        gts = graph.traversal();
+        List<Vertex> results = gts.V(sv[0]).emit().repeat(__.barrier().out("knows")).toList();
```
Comment (pluradj) [inline janusgraph-test/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java]: verify the results

Comment (twilmes) [reply janusgraph-test/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java]: This was leftover from some debugging and shouldn't have been included. I'll get rid of it as this combo of steps is covered by the TP test suite.

**Comment by pluradj** [comment]
@twilmes please squash before merging, thanks

**Comment by twilmes** [comment]
@pluradj I got them squashed. I'll merge after tests complete. I lost track of where we landed on keeping master in sync. I seem to remember merging 0.2 into master previously. Is that still the process or are we cherry picking commits?

**Pull request merged**